prov/rxm/src/rxm_cq.c

```diff
@@ -916,37 +925,59 @@ static void rxm_cq_read_write_error(struct rxm_ep *rxm_ep)
 		return;
 	}
 
-	rx_buf = (struct rxm_rx_buf *)err_entry.op_context;
+	if (RXM_IS_PROTO_STATE_TX(state)) {
+		util_cq = rxm_ep->util_ep.tx_cq;
+		util_cntr = rxm_ep->util_ep.tx_cntr;
+	}
 
-	switch (RXM_GET_PROTO_STATE(err_entry.op_context)) {
+	switch (state) {
 	case RXM_SAR_TX:
+		assert(err_entry.flags & FI_SEND);
+		sar_buf = err_entry.op_context;
+		err_entry.op_context = sar_buf->app_context;
+		err_entry.flags = ofi_tx_cq_flags(sar_buf->pkt.hdr.op);
+		break;
 	case RXM_TX:
-	case RXM_INJECT_TX:
+		assert(err_entry.flags & FI_SEND);
+		eager_buf = err_entry.op_context;
+		err_entry.op_context = eager_buf->app_context;
+		err_entry.flags = ofi_tx_cq_flags(eager_buf->pkt.hdr.op);
+		break;
 	case RXM_RNDV_TX:
-		util_cq = rxm_ep->util_ep.tx_cq;
-		if (rxm_ep->util_ep.flags & OFI_CNTR_ENABLED)
-			util_cntr = rxm_ep->util_ep.tx_cntr;
+		assert(err_entry.flags & FI_SEND);
+		rndv_buf = err_entry.op_context;
+		err_entry.op_context = rndv_buf->app_context;
+		err_entry.flags = ofi_tx_cq_flags(rndv_buf->pkt.hdr.op);
 		break;
 	case RXM_RNDV_ACK_SENT:
+		/* fall through */
 	case RXM_RX:
+		/* fall through */
 	case RXM_RNDV_READ:
+		assert(((state == RXM_RNDV_ACK_SENT) && (err_entry.flags & FI_SEND)) ||
+		       ((state == RXM_RX) && (err_entry.flags & FI_RECV)) ||
+		       ((state == RXM_RNDV_READ) && (err_entry.flags & FI_READ)));
+		rx_buf = (struct rxm_rx_buf *)err_entry.op_context;
 		util_cq = rx_buf->ep->util_ep.rx_cq;
 		util_cntr = rx_buf->ep->util_ep.rx_cntr;
+		err_entry.op_context = rx_buf->recv_entry->context;
+		err_entry.flags = rx_buf->recv_entry->comp_flags;
 		break;
 	default:
 		FI_WARN(&rxm_prov, FI_LOG_CQ, "Invalid state!\n");
 		FI_WARN(&rxm_prov, FI_LOG_CQ, "msg cq error info: %s\n",
 			fi_cq_strerror(rxm_ep->msg_cq, err_entry.prov_errno,
 				       err_entry.err_data, NULL, 0));
 		rxm_cq_write_error_all(rxm_ep, -FI_EOPBADSTATE);
-		return;
```

Comment:
Should the return be removed?  rxm_cq_write_error_all() writes errors to both CQs and all counters.

Reply:
Removed it since it was redundant. If default branch is taken util_cntr and util_cq should both be NULL and we should just return at the end of function.